### PR TITLE
Add react-testing-library

### PR DESCRIPTION
**Pull request by diegohaz**
## Description

This is a simpler alternative to #18855, and, to some extent, addresses #17249.

Comparing to #18855, we lose some features like simulating real user interactions.

`@testing-library/react` provides a `fireEvent` util, whose implementation is pretty much like this:

```js
import { act } from 'react-dom/test-utils';

const fireEvent = {
  click( element, options ) {
    return act(
      () => element.dispatchEvent( 'click', options )
    );
  },
  mouseDown() { ... },
  ...
};
```

So, calling `fireEvent.click` will only dispatch the `click` event, not the whole event stack dispatched by browsers when users perform a click action, such as `mouseDown`, `mouseUp` etc.

On the other hand, this PR is much smaller, and we don't have to maintain a new package as in #18855. If we do need those test utilities, the PR is there and we can merge it in the future. But just adding `@testing-library/react` here is already a big improvement over the current unit testing approaches we have in this project. Namely, `react-dom/test-utils`, `react-test-renderer` and `enzyme`, as demonstrated in https://github.com/WordPress/gutenberg/issues/17249#issuecomment-544337655.

As an implementation example, this PR improves the SlotFill tests that weren't covering the `bubblesVirtually` case ([`react-test-renderer` doesn't support portals](https://github.com/facebook/react/issues/11565)). This issue was mentioned on https://github.com/WordPress/gutenberg/pull/19242#issuecomment-590638543.

## How has this been tested?

`npm run test-unit`

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/slot-fill/test/__snapshots__/slot.js.snap [hunk 1]
```diff
@@ -1,128 +1,144 @@
...
-  />,
+  />
   <div
     data-position="second"
   >
     Content
-  </div>,
-]
+  </div>
+</div>
```
Comment (diegohaz) [inline packages/components/src/slot-fill/test/__snapshots__/slot.js.snap]: All the snapshots will have this wrapping `div` added by react-testing-library.

File: packages/components/src/slot-fill/test/__snapshots__/slot.js.snap [hunk 2]
```diff
@@ -1,128 +1,144 @@
...
   >
-    <div />
-  </div>,
-]
+    <div>
+      Content
+    </div>
+  </div>
+</div>
 `;
```
Comment (diegohaz) [inline packages/components/src/slot-fill/test/__snapshots__/slot.js.snap]: Now this works! :)

**Comment by ItsJonQ** [comment]
This is awesome!! Making Portals work seamlessly with testing is a big win. Not only that, but it looks like we're making tests easier to write as well with `react-testing-library` 😊

**Review by gziolo — APPROVED** [review]
Let's give it a try. It needs to be rebased :)

**Comment by gziolo** [comment]
One more thing, if this turns out to be useful, we should add RTL as a dependency for `@wordpress/jest-preset-default` and documented properly.

**Pull request merged**